src/components/utils/include/utils/timer.h

```diff
@@ -29,105 +29,72 @@
-  /**
-   * @brief constructor
-   * @param name for indentify of current timer
-   * @param task_for_tracking is SPtr to trackable task
-   */
-  Timer(const std::string& name, const TimerTask* task_for_tracking);
-
-  /**
-   * @brief destructor - if timer running : call stop in the body
-   */
+  Timer(const std::string& name, TimerTask* task);
   ~Timer();
-  /**
-   * @brief starts timer with new timeout
-   * @param timeout - time to call method from trackable class
-   * @param repeatable - should timer repeat after calling callback, or no
-   */
-  void Start(const Milliseconds timeout, const bool repeatable);
-  /**
-   * @brief stops timer without calling callback
-   */
+  
+  void Start(const Milliseconds timeout, const bool single_shot);
   void Stop();
-  /**
-   * @brief method which know about Timer state: is running or not
-   * @return true when timer runned, false when timer stand
-   */
-  bool IsRunning() const;
-  /**
-   * @brief GetTimeout
-   * @return returns timeout
-   */
-  Milliseconds GetTimeout() const;
+
+  bool is_running() const;
+  Milliseconds timeout() const;
 
  private:
+  class TimerDelegate : public threads::ThreadDelegate {
+   public:
+    explicit TimerDelegate(const Timer* timer);
+
+    void set_timeout(const Milliseconds timeout);
+    Milliseconds timeout() const;
+
+    void set_stop_flag(const bool stop_flag);
+    bool stop_flag() const;
+
+    void threadMain() OVERRIDE;
+    void exitThreadMain() OVERRIDE;
+
+   private:
+    const Timer* timer_;
+    Milliseconds timeout_;
+
+    bool stop_flag_;
+    sync_primitives::Lock state_lock_;
+    mutable sync_primitives::Lock params_lock_;
+    sync_primitives::ConditionalVariable termination_condition_;
+
+    DISALLOW_COPY_AND_ASSIGN(TimerDelegate);
+  };
+
+  void OnTimeout() const;
+
   const std::string name_;
-  const TimerTask* task_;
-  bool repeatable_;
-  uint32_t timeout_ms_;
-  bool is_running_;
-  timer_t timer_;
-  mutable sync_primitives::Lock lock_;
-  sync_primitives::Lock task_lock_;
-
-  /**
-   * @brief method called from friend handler_wrapper and call run() from task.
-   */
-  void OnTimeout();
-
-  /**
-   * @brief method for setting correct timeout.
-   * @param timeout - if it`s value = 0, timeout will be setted to 1
-   * There would be no way to stop thread if timeout in lopper will be 0
-   * and if we puts to timer_create zero timeout then we get sys error(22)
-   */
-  void SetTimeoutUnsafe(const Milliseconds timeout);
-
-  /**
-   * @brief startUnsafe, stopUnsafe - methods used for correct synchronization
-   *  and must be used only with sync_primitive (auto_lock e.g.)
-   * @return true if start/stop successfull, false if unsuccessfull
-   */
-  void StartUnsafe();
-  bool StopUnsafe();
-
-  /**
-   * @brief alone function which sends to posix_timer as callee
-   * @param signal_value - structure with parameters of posix_timer callee
-   */
-  friend void HandlePosixTimer(sigval signal_value);
+  TimerTask* task_;
```

Comment:
@LuxoftAKutsan yes, timer is responsible. But our timer class manage this pointer like `auto_ptr` - receives it in constructor and destroys in destructor, so I think `auto_ptr` is redundant here. But I can add if you insist